subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java

```diff
@@ -13,40 +13,55 @@
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationWorkerRegistry.Completion;
+import org.gradle.internal.operations.BuildOperationWorkerRegistry.Operation;
+import org.gradle.internal.progress.BuildOperationDetails;
+import org.gradle.internal.progress.BuildOperationExecutor;
 import org.gradle.process.internal.health.memory.JvmMemoryStatus;
 import org.gradle.process.internal.worker.WorkerProcess;
 
 class WorkerDaemonClient implements WorkerDaemon, Stoppable {
-    private final BuildOperationWorkerRegistry buildOperationWorkerRegistry;
     private final DaemonForkOptions forkOptions;
     private final WorkerDaemonWorker workerDaemonWorker;
     private final WorkerProcess workerProcess;
+    private final BuildOperationExecutor buildOperationExecutor;
     private int uses;
 
-    public WorkerDaemonClient(BuildOperationWorkerRegistry buildOperationWorkerRegistry, DaemonForkOptions forkOptions, WorkerDaemonWorker workerDaemonWorker, WorkerProcess workerProcess) {
-        this.buildOperationWorkerRegistry = buildOperationWorkerRegistry;
+    public WorkerDaemonClient(DaemonForkOptions forkOptions, WorkerDaemonWorker workerDaemonWorker, WorkerProcess workerProcess, BuildOperationExecutor buildOperationExecutor) {
         this.forkOptions = forkOptions;
         this.workerDaemonWorker = workerDaemonWorker;
         this.workerProcess = workerProcess;
+        this.buildOperationExecutor = buildOperationExecutor;
     }
 
     @Override
-    public <T extends WorkSpec> WorkerDaemonResult execute(WorkerDaemonAction<T> action, T spec) {
-        // currently we just allow a single compilation thread at a time (per compiler daemon)
-        // one problem to solve when allowing multiple threads is how to deal with memory requirements specified by compile tasks
-        BuildOperationWorkerRegistry.Completion workerLease = buildOperationWorkerRegistry.getCurrent().operationStart();
+    public <T extends WorkSpec> DefaultWorkResult execute(final WorkerDaemonAction<T> action, final T spec, Operation parentWorkerOperation, BuildOperationExecutor.Operation parentBuildOperation) {
+        Completion workerLease = parentWorkerOperation.operationStart();
+        BuildOperationDetails buildOperation = BuildOperationDetails.displayName(action.getDescription()).parent(parentBuildOperation).build();
         try {
-            uses++;
-            return workerDaemonWorker.execute(action, spec);
+            return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                @Override
+                public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                    uses++;
+                    return workerDaemonWorker.execute(action, spec);
+                }
+            });
         } finally {
             workerLease.operationFinish();
```

Comment:
I'm referring to the fact that we are using the `workerLease` API here, even though we are already using the buildOperationExecutor (which should get a worker lease under the hood anyway).